### ISSUE TITLE
Esc stops working when in F10 mode and form invoked, and F10 pressed again

### ISSUE THREAD

**Issue by pgimeno** [minetest/minetest#3831]
Steps to reproduce:

- Prepare a mod with this code:

```lua
minetest.register_chatcommand( 'form', {
  description = "Invoke a form, to demonstrate problem with console.",
  privs = {},
  func = function(name, param)
    minetest.show_formspec(name, "test:form", "size[6,6]")
  end
});
```
- Press F10. The console opens.
- Type /form. The console disappears and the form appears.
- Click on the form and press Esc to close it.
- Press F10 again to open the console.

Result:

- The console does not open this time. The cursor is freed, though, but not if F10 is pressed once again.
- Esc stops working for bringing up the menu.

This is happening with mods that have commands that open forms. I open the console with F10 to be able to see `/help all` in full, and then invoke one of the commands while reading the instructions (e.g. in the 'mg' map generation mod, the `/villages` or `/vmap` commands). When I try again, it stops working until I exit the game and enter again.

**Comment by est31** [comment]
Can reproduce. Likely a regression from @ShadowNinja 's changes. I think this is the bug @Calinou experienced as well but didn't know how to reproduce.

**Comment by est31** [comment]
Blocker label as it is serious enough and can be fixed fairly easily (i hope).

**Comment by pgimeno** [comment]
Update: it not only happens with F10 and not only with commands that trigger dialogs. For example, logging in to minetest.org:30023 and bringing up the text input (with / or t) before the welcome dialog appears, also triggers this problem when it does.

Tested using current dev version (dc8bf4e as of this writing).

**Comment by pgimeno** [comment]
It's probably not the correct fix, but in case it helps getting on the right track, this fixes the issue for me. I don't know if it will introduce other problems.

```diff
diff --git a/src/mainmenumanager.h b/src/mainmenumanager.h
index 17133b1..e0ea210 100644
--- a/src/mainmenumanager.h
+++ b/src/mainmenumanager.h
@@ -56,8 +56,6 @@ class MainMenuManager : public IMenuManager
 			assert(*i != menu);
 		}
 
-		if(!m_stack.empty())
-			m_stack.back()->setVisible(false);
 		m_stack.push_back(menu);
 	}
 
```

**Comment by ShadowNinja** [comment]
This should fix it: http://sprunge.us/HNWB?diff

**Comment by est31** [comment]
@ShadowNinja does it fix both #3831 and #3842 ?

**Comment by pgimeno** [comment]
That patch fixes the issue for me. I also tested #3842 and indeed it fixes it as well.

**Comment by est31** [comment]
Fixed by 8c951cae5bcfa715e227d47d122cc2de45d70a63

**Comment by ShadowNinja** [comment]
8c951cae5bcfa715e227d47d122cc2de45d70a63
